feat: Convert datetime field values to system timezone

hrwX opened the pull request:
- Let me summarize first
    - **Only one timezone should be stored in the database.**
    - **This PR makes sure all the datetime values stored in database are in system timezone, making it easier to during conversion**

- **The Issue**
    - When using the system across different timezone, the system saves datetime in two different timezones.
    - This makes it difficult to track to which the datetime string from the database belongs to, System Time/User Time/Local Time.
    - This creates an issue and confusion, since the description of the datetime field is set from `sys_defaults` which changes accoring to the user timezone.

- **Scenario**
    - Lets consider Rushabh and Nabin are the users of the system.
    - Rushabh is in LA timezone and Nabin is in IST.
    - The system timezone in System Settings is set as LA
    - Now all the document creation, modified are set as per the system timezone from System Settings since now_datetime honours system timezone by fetching the cached value `frappe.cache().get_value("time_zone")`.
    - When Rushabh does his daily stuff in the framework, since his timezone is same as system timezone, his works flawlessly, since the datetime only in LA is entered into database.
    - When Nabin does his daily stuff in the framework, since his timezone is different than system timezone, all the datetime fields that Nabin selects, are in IST, and those values are stored in the database.
    - Now we have two different timezone values in the database.
    - Refer the gif where System Timezone is America/Los_Angeles and Admin's timezone is Asia/Kolkata.
    - ![Peek 2021-06-15 23-17](https://user-images.githubusercontent.com/7310479/122099682-ec68ad80-ce2f-11eb-8714-5d37b37f3cb3.gif)

- **Proposed Solution**
    - Ideally all the time in database should be in UTC makes it easier to convert.
    - But we already have a user time zone field in User settings, which will now be considered while getting `now_datetime` in js as well as while converting to user time zone and system time zone.
    - Since we already know the system timezone, it becomes easier to format the datetime to user timezone from User doctype while displaying it to the user.
    - `moment.now` always gives us the local timezone. Now considering the above scenario, if Nabin now goes to a differnt timezone say CEST, if we use `moment,now`, all the datetime values which Nabin will insert in the datetime fields will be of CEST timezone. One more effin timezone in the system.
    - By using the user time zone in User doctype, we will be making sure that we always know what timezones are being inserting into database and it'll be easier to convert from one timezone to another.
    - ![Peek 2021-06-15 22-57](https://user-images.githubusercontent.com/7310479/122100482-de675c80-ce30-11eb-98b1-44f6b83d05da.gif)

no-docs

hrwX committed:
feat: Convert datetime field values to system timezone

Co-authored-by: Sahil Khan <sahilkhan28297@gmail.com>

hrwX committed:
fix: check if frappe.sys_defaults.time_zone exists

hrwX commented:
@rmehta @netchampfaris @surajshetty3416

rmehta commented on frappe/public/js/frappe/form/controls/datetime.js:
```diff
@@ -14,10 +26,34 @@ frappe.ui.form.ControlDatetime = class ControlDatetime extends frappe.ui.form.Co
-		const { time_zone } = frappe.sys_defaults;
-		if (!this.df.hide_timezone && !frappe.datetime.is_timezone_same()) {
+		const description = this.df.description;
+		const time_zone = frappe.boot.time_zone ? frappe.boot.time_zone.user_time_zone : frappe.sys_defaults.time_zone;
```
should be a function `get_user_time_zone`

rmehta commented on frappe/public/js/frappe/utils/datetime.js:
```diff
@@ -13,33 +13,44 @@ frappe.provide("frappe.datetime");
+		return format===false ? date_obj : date_obj.format(frappe.defaultDatetimeFormat);
 	},
 
 	is_timezone_same: function() {
```
should have ideally been `is_system_time_zone`

rmehta commented on frappe/core/doctype/user/user.py:
```diff
@@ -594,6 +595,13 @@ def find_by_credentials(cls, user_name: str, password: str, validate_password: b
+		from frappe.utils import get_time_zone
+
+		if not self.time_zone:
+			frappe.msgprint(_("User Time Zone was not set, defaulting to System Time Zone."), title=_("User Time Zone"))
```
msgprint not required

rmehta reviewed (dismissed):
Some minor fixes. Have you tested this in multiple scenarios?

@surajshetty3416 this has been a long standing issue, we should fix it.

hrwX commented:
> Some minor fixes. Have you tested this in multiple scenarios?
> 
> @surajshetty3416 this has been a long standing issue, we should fix it.

- So I am still thinking whether we should store `date` and `time` in db in datetime format, that way we can easily convert date and time as per the user timezone, we can always format the value in js to display either date/time.
- What do you think ?? @rmehta @surajshetty3416

sahil28297 commented:
@rmehta @surajshetty3416 There are two solutions to the problem mentioned above by @hrwX. 

1. The solution proposed by us in this PR, wherein we store each of the datetime values in database according to the "Timezone" value in "System Settings". This ensures that the datetime values throughout the system will be stored according to System Settings. The drawback with this approach is we are not storing the "timezone info" along with the datetime field in mariadb. This means that you cannot change the timezone settings in the future once its set (coz you won't know what the previous timezone value was and you won't be able to decide on the offset value for converting between different timezones).

2. To solve the drawback of the above problem, the other solution is to store each of the datetime values in the database as UTC. This ensures that you have the offset and the timezone every time you want to convert the datetime from one timezone to another (coz everything is in UTC) even when user wants to change their own timezone in User settings or even the timezone in System Settings. Reference: [Django Timezone Implementation](https://docs.djangoproject.com/en/3.2/topics/i18n/timezones/).

**Note:** 
1. This PR only deals with timezone implementation of "datetime" fields. **"Date" or "time" fields in the system are still timezone unaware.**

2. If this PR will be merged for site that already contains some data, there is no way on knowing whether all of the datetime values are from a single timezone or not, so this PR is under the assumption that **each of the timezone values till now in the database are from a single timezone**.

hrwX commented:
Damn stale bot!

sagarvora commented:
Maybe a patch is required to convert existing times to UTC in the database? But this could become unmanageable really quickly.
I propose avoiding the patch and storing all times in the timezone set in System Settings instead.

gavindsouza commented on frappe/public/js/frappe/utils/datetime.js:
```diff
@@ -13,33 +13,44 @@ frappe.provide("frappe.datetime");
-			return true;
+	is_system_time_zone: function() {
+		if (frappe.boot.time_zone && frappe.boot.time_zone.system_time_zone && frappe.boot.time_zone.user_time_zone) {
+			return moment().tz(frappe.boot.time_zone.system_time_zone).utcOffset() === moment().tz(frappe.boot.time_zone.user_time_zone).utcOffset();
```
Can we do this while still keeping the `is_timezone_same` util? 

If there's no reason to get rid of the API, we're better off leaving it here with a deprecation warning. Let's give those using this API enough time to make the necessary changes to make their apps/scripts/customizations compatible.

<img width="1552" alt="Screenshot 2021-10-12 at 11 20 35 AM" src="https://user-images.githubusercontent.com/36654812/136909786-44861e7c-f75f-49a3-9080-3e2499143bf4.png">

hrwX replied:
Added the API again.

gavindsouza commented on frappe/core/doctype/system_settings/system_settings.js:
```diff
@@ -32,5 +32,11 @@ frappe.ui.form.on("System Settings", {
+	after_save: function(frm) {
+		if (frappe.boot.time_zone && frappe.boot.time_zone.system_time_zone !== frm.doc.time_zone) {
+			// Clear cache after saving to refresh the values of time_zone
+			frappe.ui.toolbar.clear_cache();
```
You'd want to do this after the `time_zone` value has changed, not on every save right?

hrwX replied:
Moved it to `on_update` 😓.

gavindsouza commented on frappe/core/doctype/user/user.py:
```diff
@@ -594,6 +595,12 @@ def find_by_credentials(cls, user_name: str, password: str, validate_password: b
 		return user
 
+	def set_time_zone(self):
+		from frappe.utils import get_time_zone
```
Move this import to the top of the module instead?

gavindsouza commented on frappe/boot.py:
```diff
@@ -324,3 +326,9 @@ def get_desk_settings():
+def set_time_zone(bootinfo):
+	bootinfo.time_zone = {
+		"system_time_zone": get_time_zone(),
+		"user_time_zone": bootinfo.get("user_info", {}).get(frappe.session.user, {}).get("time_zone", None) or get_time_zone()
```
```suggestion
		"system": get_time_zone(),
		"user": bootinfo.get("user_info", {}).get(frappe.session.user, {}).get("time_zone") or get_time_zone()
```
`frappe.boot.time_zone.system` is sufficient to point out that it's the Site's timezone? Similarly for user.

gavindsouza reviewed (dismissed):
So this hasn't gotten the attention it deserves earlier. Hope we can update the branch and take it ahead now 🥲

<img width="1552" alt="Screenshot 2021-10-12 at 11 29 55 AM" src="https://user-images.githubusercontent.com/36654812/136909125-09654dd0-d27e-4532-ab06-fa9185a3b3d7.png">

So changed my user's timezone to Canada/Central and created a doc, timeline and listview say it's been created "~10 hours ago" for the user. For users that share the same timezone as the site, it simply says "Now"ish.

hrwX commented:
> Maybe a patch is required to convert existing times to UTC in the database? But this could become unmanageable really quickly. I propose avoiding the patch and storing all times in the timezone set in System Settings instead.

Patching the system is one heck of a task since we don't know which datetime are user input. We might end up messing up the datetimes by assuming the timezone. Its easier to keep the values as it is and make sure further values are fixed.

@gavindsouza Will answer and fix the review when I kinda get a bit of time.

hrwX committed:
Merge branch 'develop' of github.com:frappe/frappe into multi_timezone_support

hrwX committed:
chore: rename timezone keys

hrwX committed:
chore: move timezone clear cache to on_update

gavindsouza commented:
> So this hasn't gotten the attention it deserves earlier. Hope we can update the branch and take it ahead now 🥲
> 
> <img alt="Screenshot 2021-10-12 at 11 29 55 AM" width="1552" src="https://user-images.githubusercontent.com/36654812/136909125-09654dd0-d27e-4532-ab06-fa9185a3b3d7.png">
> 
> So changed my user's timezone to Canada/Central and created a doc, timeline and listview say it's been created "~10 hours ago" for the user. For users that share the same timezone as the site, it simply says "Now"ish.

Hey, has this been fixed yet? @hrwX

sagarvora commented:
Another point I'd like to note here is that immediately once this change is merged into a production system, there will be a possibility of time mismatch, which causes issues like this:

> Error: Document has been modified after you have opened it (2021-03-22 14:22:26.441159, 2021-03-22 14:22:26.593258). Please refresh to get the latest document.

hrwX commented:
> > So changed my user's timezone to Canada/Central and created a doc, timeline and listview say it's been created "~10 hours ago" for the user. For users that share the same timezone as the site, it simply says "Now"ish.

So the timezone in system settings changes to read-only, apart from that cant do anything since
- We do not know the exact timezone of the user if not set in user doctype
- If user has timezone in user doctype, the framework picked up time from browser, so we cant be sure what time to assume.

I guess If we had to patch this, we will have to `Assume` a lot of things, which can be dangerous esp for transactional data.

hrwX committed:
Merge branch 'develop' of github.com:frappe/frappe into multi_timezone_support

hrwX commented:
> > So this hasn't gotten the attention it deserves earlier. Hope we can update the branch and take it ahead now smiling_face_with_tear
> > <img alt="Screenshot 2021-10-12 at 11 29 55 AM" width="1552" src="https://user-images.githubusercontent.com/36654812/136909125-09654dd0-d27e-4532-ab06-fa9185a3b3d7.png">
> > So changed my user's timezone to Canada/Central and created a doc, timeline and listview say it's been created "~10 hours ago" for the user. For users that share the same timezone as the site, it simply says "Now"ish.
> 
> Hey, has this been fixed yet? @hrwX

Looking into this. it seems to be broken on my local for some weird reason

hrwX commented:
> Another point I'd like to note here is that immediately once this change is merged into a production system, there will be a possibility of time mismatch, which causes issues like this:
> 
> > Error: Document has been modified after you have opened it (2021-03-22 14:22:26.441159, 2021-03-22 14:22:26.593258). Please refresh to get the latest document.

- I don't think so this will happen since the creation and update values are set as per system timezone and not as per user timezone.
- The only change this PR will do is make the system timezone mandatory, so that the timezone in the db stays the same, as well as when a user selects datetime data, internally, before setting it in doc, system will change it to system timezone.
- While the data is being displayed to the user, we convert the system_timezone data to user timezone on the frontend

hrwX committed:
Merge branch 'develop' of github.com:frappe/frappe into multi_timezone_support

surajshetty3416 committed:
Merge branch 'develop' into multi_timezone_support

hrwX committed:
Merge branch 'develop' of github.com:frappe/frappe into multi_timezone_support

hrwX committed:
chore: make system timezone readonly

hrwX committed:
Merge branch 'multi_timezone_support' of github.com:hrwx/frappe into multi_timezone_support

surajshetty3416 committed:
Merge branch 'develop' into multi_timezone_support

surajshetty3416 committed:
fix: Pretty date rendering based on user-timezone

surajshetty3416 commented:
> <img alt="Screenshot 2021-10-12 at 11 29 55 AM" width="500" src="https://user-images.githubusercontent.com/36654812/136909125-09654dd0-d27e-4532-ab06-fa9185a3b3d7.png">
> 
> So changed my user's timezone to Canada/Central and created a doc, timeline and listview say it's been created "~10 hours ago" for the user. For users that share the same timezone as the site, it simply says "Now"ish.


@hrwX, @gavindsouza The issue should be fixed with https://github.com/frappe/frappe/pull/13504/commits/67b0293e6fc20eb449c79e26f67c4eb4d9341d46

![Screenshot 2021-11-30 at 10 08 38 AM](https://user-images.githubusercontent.com/13928957/144001565-b5b3414f-df50-45d1-802f-b96b1578c7b7.png)
An example of users in different timezones accessing same document

surajshetty3416 committed:
test: Fix test case for datetime

surajshetty3416 committed:
fix: Do not guess timezone for only time

- Time value should be consistent across timezones
- Only worry about timezone when dealing with datetime

surajshetty3416 committed:
fix: Remove unnecessary code

- frappe.datetime.str_to_user(value, false) takes care of tz conversion

surajshetty3416 committed:
fix: Select date of datepicker while doing set_formatted_input

- To avoid empty value when clicking datetime for first time

surajshetty3416 committed:
fix: Set now date based on system timezone

- because set_formatted_value will convert value to user timezone down the line

surajshetty3416 committed:
fix: Fix timezone conversions

- User to System & vice-versa
- Fixed infinite loop of setting datepicker value

surajshetty3416 committed:
Merge branch 'develop' into multi_timezone_support

surajshetty3416 committed:
Merge branch 'develop' into multi_timezone_support

surajshetty3416 commented:
@hrwX, @gavindsouza this PR seems complete... I pushed a few fixes to resolve bugs and a few edge cases. 

Can you test this out locally to verify if everything is working as expected? Planning to merge this soon.
